File: java/src/main/java/com/workshop/architecture/fitness/clean_architecture/entity/MembershipInvoiceDetails.java
package com.workshop.architecture.fitness.clean_architecture.entity;

import org.jspecify.annotations.NonNull;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.UUID;

public record MembershipInvoiceDetails(UUID id, UUID customerId, UUID membershipId,
                                       UUID planId, LocalDate dueDate, String planTitle, BigDecimal planPrice) {

    public static final int INVOICE_DUE_DAYS = 30;

    public static @NonNull MembershipInvoiceDetails create(UUID customerId, Membership membership, Plan plan) {
        var invoiceDueDate = LocalDate.now().plusDays(INVOICE_DUE_DAYS);

        return new MembershipInvoiceDetails(
                UUID.randomUUID(),
                customerId,
                membership.id(),
                plan.id(),
                invoiceDueDate,
                plan.title(),
                plan.price());
    }
}

File: python/src/workshop_api/fitness/clean_architecture/entity/membership_invoice_details.py
import uuid
from dataclasses import dataclass
from datetime import date, timedelta
from decimal import Decimal
from uuid import UUID

from workshop_api.fitness.clean_architecture.entity.membership import Membership
from workshop_api.fitness.clean_architecture.entity.plan import Plan


@dataclass(frozen=True)
class MembershipInvoiceDetails:
    id: UUID
    customer_id: UUID
    membership_id: UUID
    plan_id: UUID
    due_date: date
    plan_title: str
    plan_price: Decimal

    INVOICE_DUE_DAYS = 30

    @staticmethod
    def create(
        customer_id: UUID,
        membership: Membership,
        plan: Plan,
    ) -> MembershipInvoiceDetails:
        return MembershipInvoiceDetails(
            id=uuid.uuid4(),
            customer_id=customer_id,
            membership_id=membership.id,
            plan_id=plan.id,
            due_date=date.today() + timedelta(days=MembershipInvoiceDetails.INVOICE_DUE_DAYS),
            plan_title=plan.title,
            plan_price=plan.price,
        )
